python/pylibvw.cc

```diff
@@ -210,6 +220,145 @@ class OptionManager
   }
 };
 
+class PyCppCallback
+{
+private:
+  void* base_learner;
+
+  bool isMulti = false;
+  multi_ex* examples = nullptr;
+
+  io_buf* model_file = nullptr;
+
+public:
+  PyCppCallback(void* base_learner) : base_learner(base_learner) {}
+  PyCppCallback(io_buf* model_file) : model_file(model_file) {}
+  PyCppCallback(void* base_learner, multi_ex* examples) : base_learner(base_learner), examples(examples)
+  {
+    isMulti = true;
+  }
+
+  void CallBaseLearner(example_ptr ec, bool should_call_learn = true)
+  {
+    if (!isMulti)
+    {
+      if (should_call_learn)
+        reinterpret_cast<VW::LEARNER::single_learner*>(this->base_learner)->learn(*ec.get());
```

Comment:
Once `this->base_learner` is a `base_learner` you can use `as_singleline()`

Reply:
good idea